perfschema/statement.go

```diff
@@ -0,0 +1,322 @@
+	// Metric, number of select scan
+	selectScan uint32
+	// Metric, number of sort merge passes
+	sortMergePasses uint32
+	// Metric, number of sort merge
+	sortRange uint32
+	// Metric, number of sort rows
+	sortRows uint32
+	// Metric, number of sort scans
+	sortScan uint32
+	// Metric, no index used flag
+	noIndexUsed uint8
+	// Metric, no good index used flag
+	noGoodIndexUsed uint8
+}
+
+const (
+	// Maximum allowed number of elements in table events_statements_history.
+	// TODO: make it configurable?
+	stmtsHistoryElemMax uint64 = 1024
+)
+
+var (
+	stmtInfos = make(map[reflect.Type]*statementInfo)
+)
+
+func (ps *perfSchema) RegisterStatement(category, name string, elem interface{}) {
+	instrumentName := fmt.Sprintf("%s%s/%s", statementInstrumentPrefix, category, name)
+	key, err := ps.addInstrument(instrumentName)
+	if err != nil {
+		// just ignore, do nothing else.
+		return
+	}
+
+	stmtInfos[reflect.TypeOf(elem)] = &statementInfo{
+		key:  key,
+		name: instrumentName,
+	}
+}
+
+func (ps *perfSchema) StartStatement(sql string, connID uint64, callerName EnumCallerName, elem interface{}) *StatementState {
+	stmtType := reflect.TypeOf(elem)
+	info, ok := stmtInfos[stmtType]
+	if !ok {
+		// just ignore, do nothing else.
+		return nil
+	}
+
+	// check and apply the configuration parameter in table setup_timers.
+	timerName, err := ps.getTimerName(flagStatement)
+	if err != nil {
+		// just ignore, do nothing else.
+		return nil
+	}
+	var timerStart int64
+	switch timerName {
+	case timerNameNanosec:
+		timerStart = time.Now().UnixNano()
+	case timerNameMicrosec:
+		timerStart = time.Now().UnixNano() / int64(time.Microsecond)
+	case timerNameMillisec:
+		timerStart = time.Now().UnixNano() / int64(time.Millisecond)
+	default:
+		return nil
+	}
+
+	// TODO: check and apply the additional configuration parameters in:
+	// - table setup_actors
+	// - table setup_setup_consumers
+	// - table setup_instruments
+	// - table setup_objects
+
+	var source string
+	source, ok = callerNames[callerName]
+	if !ok {
+		_, fileName, fileLine, ok := runtime.Caller(1)
+		if !ok {
+			// just ignore, do nothing else.
+			return nil
+		}
+		source = fmt.Sprintf("%s:%d", fileName, fileLine)
+		callerNames[callerName] = source
+	}
+
+	return &StatementState{
+		connID:     connID,
+		info:       info,
+		stmtType:   stmtType,
+		source:     source,
+		timerName:  timerName,
+		timerStart: timerStart,
+		sqlText:    sql,
+	}
+}
+
+func (ps *perfSchema) EndStatement(state *StatementState) {
+	if state == nil {
+		// just ignore, do nothing else.
+		return
+	}
+
+	switch state.timerName {
+	case timerNameNanosec:
+		state.timerEnd = time.Now().UnixNano()
+	case timerNameMicrosec:
+		state.timerEnd = time.Now().UnixNano() / int64(time.Microsecond)
+	case timerNameMillisec:
+		state.timerEnd = time.Now().UnixNano() / int64(time.Millisecond)
+	default:
+		return
+	}
+
+	record := state2Record(state)
+	_ = ps.updateEventsStmtsCurrent(state.connID, record)
```

Comment:
I can revise it, the reason I don't process the value is because the design principle, performance instrumentation should not affect normal operations in any circumstances.